pages/intelligence.md

```diff
@@ -0,0 +1,137 @@
+# Code intelligence
+
+We'd like to replace elixir_sense with our own implmentation for the following bits
+of code intelligence.
+
+ * Go to definition
+ * Find references
+ * Refactoring
+
+In order to have this level of intelligence of elixir code, we need to
+search both beam files and source code to discover relevant parts and
+locations of elixir and erlang code, index them an make them
+searchable. This is a large task that will not be accomplished in one
+commit, and this document will help to lay out the scope of the work,
+as well as collect thoughts and drive the implementation.
+
+Code intelligence will be oragnized into the following discrete areas:
+
+* **Discovery** - Finding points of interest in the project's codebase, dependencies and elixir core modules
+* **Caching** - Writing the discovered items to disk and ensuring the index is up to date with the state of the project at *all times*. The goal here is to never have to delete the `.lexical` directory.
+* **Indexing** - Storing the discovered artifacts in a local in-memory index that enable fast searching.
+* **Searching** ideally using something like the [sublime text's fuzzy search](https://www.forrestthewoods.com/blog/reverse_engineering_sublime_texts_fuzzy_match/), but we definitely need some type of fuzzy search.
+* **Recognition** For several of the above examples (go to definition, find references), we need to be able to recognize what our cursor is hovering over. Is it a module? An atom? a variable? etc.
+
+# Discovery
+Lexical needs a flexible, expandable discovery engine. It needs to be
+able to examine both artifacts on disk like beam files and source
+code, and to index code that's being edited by the user. Right from
+the start, it needs to be able to find the following items:
+
+ * Modules
+ * Public and private module functions
+ * Macros
+ * Anonymous functions
+ * Structs
+ * Records
+ * Types
+ * Variables
+ * Atoms?
+ * Additional items, defined in plugins
+
+While investigating the above items, we will need to take into account
+the difference between usages and definitions. This will be pretty
+clear for most items (though aliasing and importing might muddy the
+waters somewhat), but atoms will be difficult, since there's no
+defintion to speak of, just a usage.
```

Comment:
yes, atoms are like an integer or any other primitive, the integer value might be used _somewhere_ as an indication of state, but it is just an integer like any other.

Reply:
right, but the wrinkle is that modules are atoms too

Reply:
Right! So they need to be treated like... Enums?  
A dictionary of values that are indexed as they are "found throughout the code". Perhaps it gets a little bit tricky to remove indexed atoms once they are deleted from the source.